benchmarks/src/jmh/kotlin/benchmarks/flow/TakeBenchmark.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package benchmarks.flow
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+import org.openjdk.jmh.annotations.*
+import java.util.concurrent.*
+import java.util.concurrent.CancellationException
+import kotlin.coroutines.*
+import kotlin.coroutines.intrinsics.*
+import benchmarks.flow.scrabble.flow as unsafeFlow
+
+@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
+@Fork(value = 1)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@State(Scope.Benchmark)
+open class TakeBenchmark {
+    @Param("1", "10", "100", "1000")
+    private var size: Int = 0
+
+    private suspend inline fun Flow<Long>.consume() =
+        filter { it % 2L != 0L }
+            .map { it * it }.count()
+
+    @Benchmark
+    fun baseline() = runBlocking<Int> {
+        (0L until size).asFlow().consume()
+    }
+
+    @Benchmark
+    fun originalTake() = runBlocking<Int> {
+        (0L..Long.MAX_VALUE).asFlow().originalTake(size).consume()
+    }
+
+    @Benchmark
+    fun fastPathTake() = runBlocking<Int> {
```

Comment:
Both `fastPathTake` and `mergedStateMachineTake` yield (mostly) the same performance characteristics.
In theory, merged SM can be faster because it avoids int boxing and (potentially) can avoid allocating `emitFun`, though it will complicate things (especially because this approach is not multiplatform), so I've decided to stick with `fastPath`

Reply:
Good.